### DISCUSSION TITLE
GriefPrevention and Towny

### DISCUSSION THREAD

**Discussion by MirceaDumitrescu** [GriefPrevention/GriefPrevention#10] [Ideas]
Is there any way to have both plugins active at the same time without conflicting? Is there any integration at all ? Basicly anyone could purchase town or place a chest in a already claimed land by the other plugin.

**Comment by RoboMWM** [comment]
Towny _could_ hook into GriefPrevention and prevent this behavior.

Assuming they won't do that, someone could develop an addon which makes these two work together. I do not know Towny's API (and idk if they fixed their build process), but I could help contribute to the GP side of the addon if someone does take on the project.

**Comment by Jikoo** [comment]
Had a look into this because Towny's completely undocumented and I've messed with it before. The TownClaimEvent doesn't implement Cancellable, so it gets a little hacky.
"Preventing" Towny claiming inside GP claims:

``` java
    @EventHandler
    public void onTownClaim(TownClaimEvent event) {
        final TownBlock townBlock = event.getTownBlock();
        for (Claim claim : GriefPrevention.instance.dataStore.getClaims(townBlock.getX(), townBlock.getZ())) {
            // Ensure claims are in the same world
            if (claim.getGreaterBoundaryCorner().getWorld().getName().equals(townBlock.getWorld().getName())) {
                // Remove chunk from town later
                new BukkitRunnable() {
                    @Override
                    public void run() {
                        townBlock.getTown().removeTownBlock(townBlock);
                    }
                }.runTask(plugin);
                return;
            }
        }
    }
```

Unfortunately, GP does not appear to offer any events when creating/resizing claims at all, so this either has to be done on GP's side or GP needs the relevant events (ideally, cancellable) added.

Honestly, if the events needed are to be added and Towny integration is not, GP should probably drop out-of-the-box WG support and instead offer an addon plugin with various hooks.

Were a ClaimCreateEvent added, the following code should work for the other side:

``` java
    @EventHandler
    public void onClaimCreate(ClaimCreateEvent event) {
        Claim claim = event.getClaim();
        Location lesser = claim.getLesserBoundaryCorner();
        Location greater = claim.getGreaterBoundaryCorner();
        int minChunkX = lesser.getBlockX() >> 4;
        int maxChunkX = greater.getBlockX() >> 4;
        int minChunkZ = lesser.getBlockZ() >> 4;
        int maxChunkZ = greater.getBlockZ() >> 4;
        String world = lesser.getWorld().getName();
        for (int x = minChunkX; x <= maxChunkX; x++) {
            for (int z = minChunkZ; z <= maxChunkZ; z++) {
                // Check each chunk inside the claim to ensure it's not part of a Town
                try {
                    if (new WorldCoord(world, x, z).getTownBlock().hasTown()) {
                        event.setCancelled(true);
                        return;
                    }
                } catch (Exception e) {
                    // NotRegisteredException is thrown when a WorldCoord is not registered to a Town/wilderness, do nothing
                }
            }
        }
    }
```

Denial messages would be nice, but in terms of raw functionality, that should do. If you're looking to offer maximum support for plugins, some way to get the player (if any) doing the claiming/resizing would be nice for that reason. I'd imagine on the Towny side it'd be as simple as warning all players with access to expanding the Town in the chunk that a GP claim is present, though I'm not at all familiar with Towny's player handling.

Edit: Oh, and Towny's still built with an Ant script that downloads its dependencies off of Towny's site.

**Comment by RoboMWM** [comment]
Yes, definitely looking into making these all events. I wouldn't mind dropping WorldGuard support.

Might also be worth looking into Blockshub, where such a bridge plugin has been added.

**Comment by LlmDl** [comment]
For what's it's worth now, we have a PreTownClaimEvent which is cancelable that could be used to stop Towny claims being done on GP claims.
